File: zgiis/navigation/national_gnss_status.py
"""Zimbabwe National GNSS Status block — live city accuracy + services readout."""
from __future__ import annotations

from typing import Any

from zgiis.navigation.gnss_forecast import (
    FORECAST_SITES,
    ForecastStatus,
    GnssForecastCity,
    _expected_accuracy,
)

NATIONAL_GNSS_CITY_ORDER = [site["city"] for site in FORECAST_SITES]

_CITY_DISPLAY = {site["city"]: site["displayName"] for site in FORECAST_SITES}

_STATUS_EMOJI = {"excellent": "🟢", "moderate": "🟡", "warning": "🔴"}

_NATIONAL_SERVICES: dict[ForecastStatus, dict[str, str]] = {
    "excellent": {
        "Surveying": "Available",
        "Agriculture": "Excellent",
        "Mining": "Good",
        "Aviation": "Normal",
        "Power Grid": "Low Risk",
    },
    "moderate": {
        "Surveying": "Advisory",
        "Agriculture": "Excellent",
        "Mining": "Good",
        "Aviation": "Advisory",
        "Power Grid": "Low Risk",
    },
    "warning": {
        "Surveying": "Limited",
        "Agriculture": "Monitor",
        "Mining": "Caution",
        "Aviation": "Advisory",
        "Power Grid": "Moderate Risk",
    },
}


def storm_risk_label(sw: dict[str, Any] | None) -> str:
    if not sw or sw.get("kp") is None:
        return "Updating"
    kp = float(sw["kp"])
    if kp >= 7:
        return "Severe"
    if kp >= 5:
        return "High"
    if kp >= 4:
        return "Moderate"
    if kp >= 3:
        return "Unsettled"
    return "Low"


def _field(fc: GnssForecastCity, label: str) -> str | None:
    for f in fc.fields:
        if f.get("label") == label:
            return f.get("value")
    return None


def _city_accuracy_line(fc: GnssForecastCity) -> tuple[str, str]:
    emoji = _STATUS_EMOJI[fc.status]
    accuracy = _field(fc, "Expected Accuracy")
    if accuracy:
        return emoji, accuracy
    iono = fc.iono_stress if fc.iono_stress is not None else 50.0
    feed = fc.feed_reliability if fc.feed_reliability is not None else 50.0
    return emoji, _expected_accuracy(iono, feed)


def build_national_gnss_status_block(
    forecasts: list[GnssForecastCity],
    tone: ForecastStatus,
    sw: dict[str, Any] | None,
) -> str:
    """Return the 🇿🇼 ZIMBABWE NATIONAL GNSS STATUS block from live forecast inputs."""
    by_city = {f.city: f for f in forecasts}
    city_lines: list[str] = []
    for city_key in NATIONAL_GNSS_CITY_ORDER:
        fc = by_city.get(city_key)
        if fc is None:
            continue
        emoji, value = _city_accuracy_line(fc)
        name = _CITY_DISPLAY.get(city_key, city_key.title())
        city_lines.append(f"{emoji} {name}\n{value}")

    kp = sw.get("kp") if sw else None
    kp_line = f"Kp = {int(kp) if kp is not None and kp == int(kp) else kp}" if kp is not None else "Kp = Updating"
    storm_line = f"Storm Risk = {storm_risk_label(sw)}"
    services = _NATIONAL_SERVICES[tone]

    lines = [
        "🇿🇼 ZIMBABWE NATIONAL GNSS STATUS",
        "",
        "GNSS Accuracy Today",
        *city_lines,
        "",
        "Current Space Weather",
        kp_line,
        storm_line,
        "",
        "National Services",
    ]
    for service, label in services.items():
        lines.append(f"{service}\n{label}")

    return "\n".join(lines)

File: frontend/lib/nationalGnssStatus.ts
import type { ForecastStatus, GnssForecastCity } from "./gnssWeatherIntelligence";
import { expectedAccuracy } from "./gnssForecastEngine";
import type { SpaceWeatherCurrent } from "./types";

export const NATIONAL_GNSS_CITY_ORDER = ["HARARE", "BULAWAYO", "MUTARE", "VICTORIA FALLS"] as const;

const CITY_DISPLAY: Record<string, string> = {
  HARARE: "Harare",
  BULAWAYO: "Bulawayo",
  MUTARE: "Mutare",
  "VICTORIA FALLS": "Victoria Falls",
};

const STATUS_EMOJI: Record<ForecastStatus, string> = {
  excellent: "🟢",
  moderate: "🟡",
  warning: "🔴",
};

const NATIONAL_SERVICES: Record<ForecastStatus, Record<string, string>> = {
  excellent: {
    Surveying: "GNSS NORMAL",
    Agriculture: "GNSS NORMAL",
    Mining: "GNSS NORMAL",
    Aviation: "GNSS NORMAL",
    "Power Grid": "LOW RISK",
  },
  moderate: {
    Surveying: "GNSS ADVISORY",
    Agriculture: "GNSS ADVISORY",
    Mining: "GNSS ADVISORY",
    Aviation: "GNSS ADVISORY",
    "Power Grid": "LOW–MODERATE",
  },
  warning: {
    Surveying: "GNSS CAUTION",
    Agriculture: "GNSS CAUTION",
    Mining: "GNSS CAUTION",
    Aviation: "GNSS CAUTION",
    "Power Grid": "MODERATE RISK",
  },
};

/** Maps Navigation News audience cards to National Services sector names. */
export const AUDIENCE_NATIONAL_SERVICE: Record<string, string> = {
  surveyors: "Surveying",
  farmers: "Agriculture",
  pilots: "Aviation",
  power: "Power Grid",
  telecom: "Telecommunications",
};

const TELECOM_SERVICE_STATUS: Record<ForecastStatus, string> = {
  excellent: "GNSS NORMAL",
  moderate: "GNSS ADVISORY",
  warning: "GNSS CAUTION",
};

const TONE_RANK: Record<ForecastStatus, number> = {
  excellent: 0,
  moderate: 1,
  warning: 2,
};

function worseTone(a: ForecastStatus, b: ForecastStatus): ForecastStatus {
  return TONE_RANK[a] >= TONE_RANK[b] ? a : b;
}

export function nationalServiceStatusColor(label: string, tone: ForecastStatus): string {
  if (
    label === "GNSS CAUTION" ||
    label === "MODERATE RISK" ||
    label === "Limited" ||
    label === "Caution"
  ) {
    return "#ef4444";
  }
  if (label === "Monitor" || label === "LOW–MODERATE") return tone === "warning" ? "#ef4444" : "#eab308";
  if (label === "GNSS ADVISORY" || label === "Advisory") return "#eab308";
  if (tone === "warning") return "#ef4444";
  return "#00ff88";
}

export function getAudienceNationalServiceStatus(
  audienceId: string,
  tone: ForecastStatus,
  sectorTone?: ForecastStatus,
): { serviceName: string; statusLabel: string; effectiveTone: ForecastStatus } | null {
  const serviceName = AUDIENCE_NATIONAL_SERVICE[audienceId];
  if (!serviceName) return null;
  const effectiveTone = sectorTone ? worseTone(tone, sectorTone) : tone;
  if (audienceId === "telecom") {
    return { serviceName, statusLabel: TELECOM_SERVICE_STATUS[effectiveTone], effectiveTone };
  }
  const statusLabel = NATIONAL_SERVICES[effectiveTone][serviceName];
  if (!statusLabel) return null;
  return { serviceName, statusLabel, effectiveTone };
}

function forecastField(fc: GnssForecastCity, label: string): string | undefined {
  return fc.fields.find((f) => f.label === label)?.value;
}

/** Same Expected Accuracy ranges used in Navigation News forecast cards and surveyor briefs. */
function cityAccuracyLine(fc: GnssForecastCity): { emoji: string; value: string } {
  const fromField = forecastField(fc, "Expected Accuracy");
  const iono = fc.ionoStress ?? 50;
  const feed = fc.feedReliability ?? 50;
  return {
    emoji: STATUS_EMOJI[fc.status],
    value: fromField ?? expectedAccuracy(iono, feed),
  };
}

export function stormRiskLabel(sw: SpaceWeatherCurrent | null): string {
  if (sw?.kp == null) return "Updating";
  const kp = sw.kp;
  if (kp >= 9) return "G5 Extreme";
  if (kp >= 8) return "G4 Severe";
  if (kp >= 7) return "G3 Strong";
  if (kp >= 6) return "G2 Moderate";
  if (kp >= 5) return "G1 Minor Storm";
  if (kp >= 4) return "G0 Active (below G1)";
  if (kp >= 3) return "G0 Unsettled";
  return "G0 Quiet";
}

export interface NationalGnssStatusData {
  title: string;
  cities: { name: string; emoji: string; value: string }[];
  kp: string;
  stormRisk: string;
  services: { name: string; label: string; color: string }[];
}

export function buildNationalGnssStatusData(
  forecasts: GnssForecastCity[],
  tone: ForecastStatus,
  sw: SpaceWeatherCurrent | null,
): NationalGnssStatusData {
  const byCity = Object.fromEntries(forecasts.map((f) => [f.city, f]));
  const cities = NATIONAL_GNSS_CITY_ORDER.flatMap((cityKey) => {
    const fc = byCity[cityKey];
    if (!fc) return [];
    const { emoji, value } = cityAccuracyLine(fc);
    return [{ name: CITY_DISPLAY[cityKey] ?? cityKey, emoji, value }];
  });

  const kp =
    sw?.kp != null
      ? `Kp = ${Number.isInteger(sw.kp) ? sw.kp : sw.kp}`
      : "Kp = Updating";

  const services = Object.entries(NATIONAL_SERVICES[tone]).map(([name, label]) => ({
    name,
    label,
    color: nationalServiceStatusColor(label, tone),
  }));

  return {
    title: "🇿🇼 ZIMBABWE NATIONAL GNSS STATUS",
    cities,
    kp,
    stormRisk: `Geomagnetic storm = ${stormRiskLabel(sw)}`,
    services,
  };
}

export function buildNationalGnssStatusBlock(
  forecasts: GnssForecastCity[],
  tone: ForecastStatus,
  sw: SpaceWeatherCurrent | null,
): string {
  const data = buildNationalGnssStatusData(forecasts, tone, sw);
  const cityLines = data.cities.flatMap((c) => [`${c.emoji} ${c.name}`, c.value]);
  const serviceLines = data.services.flatMap((s) => [s.name, s.label]);

  return [
    data.title,
    "",
    "GNSS Accuracy Today",
    ...cityLines,
    "",
    "Current Space Weather",
    data.kp,
    data.stormRisk,
    "",
    "National Services",
    ...serviceLines,
  ].join("\n");
}
